plenum/test/node_request/test_unit_setup_for_non_master.py

```diff
@@ -0,0 +1,168 @@
+    for node in txnPoolNodeSet:
+        replica.preparesWaitingForPrePrepare[replica.viewNo, ppSeqNo] \
+            .append((prepare, node.name))
+    replica._setup_last_ordered_for_non_master()
+    assert replica.last_ordered_3pc == (0, 0)
+
+
+def test_setup_last_ordered_for_non_master_without_prepare(txnPoolNodeSet,
+                                                           sdk_wallet_client):
+    inst_id = 1
+    replica = getNonPrimaryReplicas(txnPoolNodeSet, inst_id)[-1]
+    replica.preparesWaitingForPrePrepare.clear()
+    replica.prePreparesPendingPrevPP.clear()
+    replica.last_ordered_3pc = (0, 0)
+    timestamp = time.time()
+    ppSeqNo = 5
+    preprepare, prepare = \
+        _create_prepare_and_preprepare(inst_id,
+                                       replica.viewNo,
+                                       ppSeqNo,
+                                       timestamp,
+                                       sdk_wallet_client)
+    replica.prePreparesPendingPrevPP[replica.viewNo, ppSeqNo] = deque()
+    replica.prePreparesPendingPrevPP[replica.viewNo, ppSeqNo] \
+        .append((preprepare, replica.primaryName))
+    replica.preparesWaitingForPrePrepare[replica.viewNo, ppSeqNo] = deque()
+    replica.preparesWaitingForPrePrepare[replica.viewNo, ppSeqNo] \
+        .append((prepare, txnPoolNodeSet[-1].name))
+    replica._setup_last_ordered_for_non_master()
+    assert replica.last_ordered_3pc == (0, 0)
+
+
+def test_setup_last_ordered_for_non_master_for_master(txnPoolNodeSet,
+                                                      sdk_wallet_client):
+    inst_id = 0
+    replica = getNonPrimaryReplicas(txnPoolNodeSet, inst_id)[-1]
+    replica.preparesWaitingForPrePrepare.clear()
+    replica.prePreparesPendingPrevPP.clear()
+    replica.last_ordered_3pc = (0, 0)
+    timestamp = time.time()
+    ppSeqNo = 5
+    preprepare, prepare = \
+        _create_prepare_and_preprepare(inst_id,
+                                       replica.viewNo,
+                                       ppSeqNo,
+                                       timestamp,
+                                       sdk_wallet_client)
+    replica.prePreparesPendingPrevPP[replica.viewNo, ppSeqNo] = deque()
+    replica.prePreparesPendingPrevPP[replica.viewNo, ppSeqNo] \
+        .append((preprepare, replica.primaryName))
+    replica.preparesWaitingForPrePrepare[replica.viewNo, ppSeqNo] = deque()
+    for node in txnPoolNodeSet:
+        replica.preparesWaitingForPrePrepare[replica.viewNo, ppSeqNo] \
+            .append((prepare, node.name))
+    replica._setup_last_ordered_for_non_master()
+    assert replica.last_ordered_3pc == (0, 0)
+
+
+def test_setup_last_ordered_for_non_master_without_catchup(txnPoolNodeSet,
+                                                           sdk_wallet_client):
+    inst_id = 1
+    last_ordered_3pc = (5, 12)
+    timestamp = time.time()
+    ppSeqNo = 13
```

Comment:
There must be a gap in received PrePrepares to make conditions where requesting missed 3PC-messages may potentially occur.

Reply:
It's only unit test for _setup_last_ordered_for_non_master(). It can't requesting missed 3PC-messages